cpp/src/arrow/util/stl_util_test.cc

```diff
@@ -92,5 +93,66 @@ TEST(StlUtilTest, ArgSortPermute) {
   ExpectSortPermutation({b, c, d, e, a, f}, {4, 0, 1, 2, 3, 5}, 2);
 }
 
+TEST(StlUtilTest, VectorFlatten) {
+  std::vector<int> a{1, 2, 3};
+  std::vector<int> b{4, 5, 6};
+  std::vector<int> c{7, 8, 9};
+  std::vector<std::vector<int>> vecs{a, b, c};
+  auto actual = FlattenVectors(vecs);
+  std::vector<int> expected{1, 2, 3, 4, 5, 6, 7, 8, 9};
+  ASSERT_EQ(expected, actual);
+}
+
+TEST(StlUtilTest, VectorMap) {
+  std::vector<int> input{1, 2, 3};
+  std::vector<std::string> expected{"1", "2", "3"};
+  auto actual = MapVector([](int item) { return std::to_string(item); }, input);
```

Comment:
For the record, is it possible to write directly `MapVector(std::to_string, input)`?

Reply:
No.  There is a comment in `call_traits`(which I use to determine the return type):

`/// TODO(bkietz) support function pointers`

I tried a few `decltype` alternatives just now and couldn't make any headway.

Reply:
How about:
```c++
template <typename Fn, typename From, typename To=decltype(std::declval<Fn>()(std::declval<From>()))>
std::vector<To> MapVector(Fn&& map, const std::vector<From>& source) {
```

Reply:
Much better (I was doing `Fn(std::declval<From>)` and the errors confused me but now I see why that was incorrect).  It still won't work with `std::to_string` because the function is overloaded but it works with other function pointers now.

Reply:
Hmm, I'm a bit surprised that is doesn't work with an overloaded function, because the point of using `decltype` of an actual expression is to resolve the right overload.

Reply:
Ok, I see: `Fn` can't be resolved if the function is overloaded. Too bad.

Reply:
To be fair the STL didn't solve this case either, you cannot do `std::bind(std::to_string, 7);`.  So I think your approach is as good as it gets :).  If you're determined you can do `MapVector(static_cast<std::string (*)(int)>(std::to_string), input);`